### PR TITLE
[FLINK-8302][table]Add SHIFT_LEFT and SHIFT_RIGHT supported in SQL

### PR DESCRIPTION
**Pull request by dubin555**
In this PR, SHIFT_LEFT and SHIFT_RIGHT are added.

## General
The pull request references the related JIRA issue, [FLINK-8302]Support shift_left and shift_right in TableAPI
The pull request addresses only one issue
Each commit in the PR has a meaningful commit message (including the JIRA id)

## Tests & Build
Functionality added by the pull request is covered by tests

## Documentation
Does this pull request introduce a new feature? (yes )
If yes, how is the feature documented? (docs / JavaDocs)

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala [hunk 1]
```diff
@@ -58,4 +58,22 @@ object ScalarSqlFunctions {
       OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC)),
     SqlFunctionCategory.NUMERIC)
 
+  val SHIFT_LEFT = new SqlFunction(
+    "SHIFT_LEFT",
+    SqlKind.OTHER_FUNCTION,
+    ReturnTypes.ARG0,
+    null,
+    OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC),
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala]: Second operand should be  SqlTypeFamily.INTEGER.
 `OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC) -> OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.INTEGER) `

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala [hunk 2]
```diff
@@ -58,4 +58,22 @@ object ScalarSqlFunctions {
...
+    SqlFunctionCategory.NUMERIC
+  )
+
+  val SHIFT_RIGHT = new SqlFunction(
+    "SHIFT_RIGHT",
+    SqlKind.OTHER_FUNCTION,
+    ReturnTypes.ARG0,
+    null,
+    OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC),
+    SqlFunctionCategory.NUMERIC
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala]: Second operand should be  SqlTypeFamily.INTEGER.
 `OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC) -> OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.INTEGER) `

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 3]
```diff
@@ -1216,6 +1216,69 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+    testSqlApi(
+      "SHIFT_LEFT(21,1)",
+      "42"
+    )
+
+    testSqlApi(
+      "SHIFT_LEFT(21,1)",
+      "42"
+    )
+
```
[the remark this answers is not in the document]
Comment (dubin555) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: removed

**Comment by dubin555** [comment]
Hi @sunjincheng121 . Thanks for the review.
I already finished some changed for code cleaning and remove the duplicate test case base on your comment.

There are some similar implements in other Database, 
like MySQL, https://dev.mysql.com/doc/refman/5.7/en/bit-functions.html
Oracle, https://docs.oracle.com/cd/E22583_01/DR/help/BitShift.html
In Apache Spark, also exists similar implements, https://issues.apache.org/jira/browse/SPARK-8223

Thanks,
Du Bin

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 4]
```diff
@@ -1216,6 +1216,64 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+    testSqlApi(
+      "SHIFT_LEFT(9223372036854775807,-2147483648)",
+      "9223372036854775807"
+    )
+  }
+
+  @Test
+  def testShiftRight(): Unit = {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: Add a test case using column defined in `ScalarTypesTestBase  `, just like as follow:  
```
testSqlApi(
      "SHIFT_RIGHT(f4,1)", // f4 is BIGINT type
      "22"
    )
```

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 5]
```diff
@@ -1216,6 +1216,64 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
     )
   }
 
+  @Test
+  def testShiftLeft(): Unit = {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: Add a test case using column defined in `ScalarTypesTestBase  `, just like as follow:  
```
 testSqlApi(
      "SHIFT_LEFT(f4,1)", // f4 is BIGINT type
      "88"
    )
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala [hunk 6]
```diff
@@ -58,4 +58,22 @@ object ScalarSqlFunctions {
       OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC)),
     SqlFunctionCategory.NUMERIC)
 
+  val SHIFT_LEFT = new SqlFunction(
+    "SHIFT_LEFT",
+    SqlKind.OTHER_FUNCTION,
+    ReturnTypes.ARG0,
+    null,
+    OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.INTEGER),
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala]: I check the type defined in calcite, `SqlTypeFamily.INTEGER` is 
 `SqlTypeName.INT_TYPES` contains `(TINYINT, SMALLINT, INTEGER, BIGINT)`, So we should change `SqlTypeFamily.NUMERIC` to `SqlTypeFamily.INTEGER` .

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala [hunk 7]
```diff
@@ -58,4 +58,22 @@ object ScalarSqlFunctions {
...
+    SqlFunctionCategory.NUMERIC
+  )
+
+  val SHIFT_RIGHT = new SqlFunction(
+    "SHIFT_RIGHT",
+    SqlKind.OTHER_FUNCTION,
+    ReturnTypes.ARG0,
+    null,
+    OperandTypes.family(SqlTypeFamily.NUMERIC, SqlTypeFamily.INTEGER),
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/sql/ScalarSqlFunctions.scala]: I check the type defined in calcite, `SqlTypeFamily.INTEGER` is 
 `SqlTypeName.INT_TYPES` contains `(TINYINT, SMALLINT, INTEGER, BIGINT)`, So we should change `SqlTypeFamily.NUMERIC` to `SqlTypeFamily.INTEGER` .

**Comment by dubin555** [comment]
Hi @sunjincheng121 , thanks for the overall review of my code. I've done some typo fix and add some test case base on your comment. Really appreciate for your careful review!

Du Bin

**Comment by sunjincheng121** [comment]
Thanks @dubin555, The PR looks pretty good.
+1 to merged.

Thanks, Jincheng

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala [hunk 8]
```diff
@@ -414,6 +414,34 @@ object FunctionGenerator {
     DOUBLE_TYPE_INFO,
     BuiltInMethods.LOG_WITH_BASE)
 
+  addSqlFunctionMethod(
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala]: Shift operations are core binary operators. I would not add them here but in `ScalarOperators`. Maybe we can already reuse `generateArithmeticOperator`.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 9]
```diff
@@ -1216,6 +1216,74 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
+    testSqlApi(
```
Comment (twalthr) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: Add tests with null arguments.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala (hunk 8 above)
Comment (sunjincheng121) [reply flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala]: Yes, that's correct..
+1 to add `shift` as a binary operator.

**Comment by twalthr** [comment]
Are you able to update this PR @dubin555? Otherwise please close this PR and unassign yourself from the issue. Maybe somebody else wants to build on top of your code and finish this. Thanks.

**Pull request closed without merging**